### PR TITLE
APM-313824 add exception throttling to non-info logs

### PR DESCRIPTION
**Pull request by equardo**
Added exception throttling for all type of inner logs except info logs - if they should be throttled as well, please let me know.
azure-throttling

### CODE REVIEW

File: logs_ingest/logging.py [hunk 1]
```diff
@@ -15,28 +15,54 @@
...
     logging.info(_version_tag + msg, *args, **kwargs)
 
 
-def debug(msg, *args, **kwargs):
-    logging.debug(_version_tag + msg, *args, **kwargs)
\ No newline at end of file
+def debug(msg, caller: str, *args, **kwargs):
+    if check_if_caller_exceeded_limit(caller):
+        return
+    logging.debug(_version_tag + msg, *args, **kwargs)
```
Comment (wojtasskorcz) [inline logs_ingest/logging.py:53]: If we're not throttling info logs, I think we shouldn't throttle debug logs either. As far as I understand, debug and info logs are more "predictable" (they're not used in exceptional situations) and so they don't require throttling?

File: logs_ingest/logging.py [hunk 2]
```diff
@@ -15,28 +15,54 @@
...
+    log_calls_left = LOG_THROTTLING_LIMIT_PER_CALLER - log_call_count.get(caller, 0)
+
+    if log_calls_left == 0:
+        log_call_count[caller] = log_call_count.get(caller, 0) + 1
+        logging.warning(_version_tag + f"Logging calls from caller '{caller}' exceeded the throttling limit of"
+                        f" {LOG_THROTTLING_LIMIT_PER_CALLER}. Further logs from this caller will be discarded")
+
+    caller_exceeded_limit = log_calls_left <= 0
+    if not caller_exceeded_limit:
+        log_call_count[caller] = log_call_count.get(caller, 0) + 1
```
Comment (wojtasskorcz) [inline logs_ingest/logging.py:66]: The same value is extracted from the dict three times. Maybe we should just extract it once at the beginning of the function and assign it to a variable, e.g. `log_calls_performed`?

Comment (equardo) [reply logs_ingest/logging.py:66]: It's extracted two times at most during one execution but it's probably still better to just assign it, will do.

Comment (wojtasskorcz) [reply logs_ingest/logging.py]: Yeah, I don't mean runtime overhead here but rather code duplication

File: logs_ingest/main.py [hunk 3]
```diff
@@ -102,7 +102,7 @@ def is_too_old(timestamp: str, self_monitoring: SelfMonitoring, log_part: str):
                 return True
         except Exception:
             # Not much we can do when we can't parse the timestamp
-            logging.debug(f"Failed to parse timestamp {timestamp}")
+            logging.debug(f"Failed to parse timestamp {timestamp}", "timestamp-parsing-debug")
```
Comment (mswiatkowska) [inline logs_ingest/main.py:105]: I agree with @wojtasskorcz that we shouldn't throttle debug logs but still this log should be throttled - could you change this log to exception?

Comment (equardo) [reply logs_ingest/main.py:105]: As far as I can tell by testing using integration tests, this log is more of a regular occurrence - it happens during test_main_success multiple times:
image
This is how test_main_success execution looks like with this log on exception(). With debug() nothing is printed.

Comment (wojtasskorcz) [reply logs_ingest/main.py:105]: Either the test should be fixed (these are definitely incorrect timestamps!) or, if these timestamps are deliberately like this, we should actually log an exception. I wouldn't expect it to be a regular occurrence in production. I agree with @mswiatkowska

Comment (equardo) [reply logs_ingest/main.py]: Done, thank you

**Review by wojtasskorcz — DISMISSED** [review]

File: logs_ingest/main.py (hunk 3 above)
Comment (mswiatkowska) [reply logs_ingest/main.py]: Yes, those incorrect timestamps (for several records) are used in integration test deliberately (we check if we count parsing errors correctly)

**Review by mswiatkowska — APPROVED** [review]

File: logs_ingest/main.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def is_too_old(timestamp: str, self_monitoring: SelfMonitoring, log_part: str):
                 return True
         except Exception:
             # Not much we can do when we can't parse the timestamp
-            logging.debug(f"Failed to parse timestamp {timestamp}")
+            logging.exception(f"Failed to parse timestamp {timestamp}", "timestamp-parsing-debug")
```
Comment (mswiatkowska) [inline logs_ingest/main.py:105]: [Minor] please change timestamp-parsing-debug to timestamp-parsing-exception

Comment (equardo) [reply logs_ingest/main.py]: Fixed

**Review by mswiatkowska — APPROVED** [review]

**Review by wojtasskorcz — APPROVED** [review]

**Review by Dawid-Kaszubski — APPROVED** [review]

**Pull request merged**